{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Using the RDP Classifier to add taxonomic classification to the sequences.**  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Taxonomic classification of Diptera sequences using RDP classifier**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coleoptera_outgroup.fasta\n",
      "dipt99_headers_fasta\n",
      "diptera_clean_headers_fasta\n",
      "diptera_headers_fasta\n",
      "enafroCOI_all_clean_Diptera_95g.fasta\n",
      "enafroCOI_all_clean_Diptera_95g.fasta.reduced\n",
      "enafroCOI_all_clean_Diptera.fasta\n",
      "enafroCOI_all_clean_Diptera.fasta.reduced\n",
      "enafroCOI_all_clean_Diptera_taxa\n",
      "enafroCOI_all_clean_Diptera_taxa1\n",
      "enafroCOI_diptera_500to700_data-650to660.fasta\n",
      "enafroCOI_diptera_500to700_data.fasta\n",
      "enafroCOI_diptera_650to660_data.fasta\n",
      "enafroCOI_diptera_all_data.fasta\n",
      "enafroCOI_diptera_all_data_raw.fasta\n",
      "enafroCOI_diptera_all_data_taxa\n",
      "enafroCOI_diptera_Over499_data.fasta\n",
      "enafroCOI_diptera_Over700_data.fasta\n",
      "enafroCOI_diptera_taxa\n",
      "enafroCOI_diptera_taxa_clash.tsv\n",
      "enafroCOI_diptera_taxa.tsv\n",
      "enafroCOI_diptera_Under500_data.fasta\n",
      "genera_distro_clean_95g.txt\n",
      "genera_distro.txt\n",
      "hymenoptera_Erromyrma_outgroup.fasta\n",
      "hymenoptera_Hypoponera_outgroup.fasta\n",
      "hymenoptera_outgroup.fasta\n",
      "hymenoptera_Plagiolepis_outgroup.fasta\n",
      "hymenoptera_Tetraponera_outgroup.fasta\n",
      "lepidotera_outgroup.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_500to700_data-650to660.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_650to660.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_dgpd.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera.fasta\n",
      "Mpala_Kinondo_Turkana_Malaise_traps_dgpd_Diptera_Under500.fasta\n",
      "outgroup.fasta\n",
      "rooting_inputfile.fas\n",
      "rooting_inputfile.fas.reduced\n",
      "species_distro_clean_95g.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "\n",
    "# Executing RDP classifier, this includes the setting of the maximum size or memeory -Xmx**** to 80g\n",
    "#java -Xmx80g -jar ~/bioinformatics/github/co1_metaanalysis/code/tools/RDPTools/classifier.jar classify -c 0.8 -t ~/bioinformatics/github/co1_metaanalysis/code/tools/CO1Classifierv3.2/mydata_trained/rRNAClassifier.properties -o enafroCOI_all_clean_Diptera_taxa1 enafroCOI_all_clean_Diptera.fasta\n",
    "\n",
    "source ../../../code/classify.sh\n",
    "RDPclassifyCOI enafroCOI_all_clean_Diptera.fasta enafroCOI_diptera_all_data.fasta << EOF\n",
    "80g\n",
    "80g\n",
    "EOF\n",
    "\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Reformating the taxa file to tsv**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process_ID\tPhylum\tClass\tOrder\tFamily\tSubfamily\tTribe\tGenus\tSpecies\tSubspecies\tCountry\tExactsite\tLatitude\tLongitude\tElevation\tseq_len\tAs_CellularOrganism\tCellularOrganism_sc\tAs_superkingdom\tSuperkingdom_sc\tAs_kingdom\tKingdom_sc\tAs_phylum\tPhylum_sc\tAs_class\tClass_sc\tAs_order\tOrder_sc\tAs_family\tFamily_sc\tAs_genus\tGenus_sc\tAs_species\tSpecies_sc\n",
      "GBDP19110-15\tArthropoda\tInsecta\tDiptera\tPsychodidae\tPhlebotominae\tNA\tSergentomyia\tSergentomyia_minuta\tNA\tAlgeria\tNA\tNA\tNA\tNA\t564\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tPsychodidae\t1.0\tSergentomyia\t1.0\tSergentomyia_minuta\t1.0\n",
      "GBDP19109-15\tArthropoda\tInsecta\tDiptera\tPsychodidae\tPhlebotominae\tNA\tSergentomyia\tSergentomyia_minuta\tNA\tAlgeria\tNA\tNA\tNA\tNA\t564\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tPsychodidae\t1.0\tSergentomyia\t1.0\tSergentomyia_minuta\t1.0\n",
      "GBDP19118-15\tArthropoda\tInsecta\tDiptera\tPsychodidae\tPhlebotominae\tNA\tSergentomyia\tSergentomyia_minuta\tNA\tAlgeria\tNA\tNA\tNA\tNA\t564\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tPsychodidae\t1.0\tSergentomyia\t1.0\tSergentomyia_minuta\t1.0\n",
      "GBDP4645-08\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_imicola\tNA\tAlgeria\tNA\tNA\tNA\tNA\t471\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_imicola\t1.0\n",
      "GBDP4646-08\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_imicola\tNA\tAlgeria\tNA\tNA\tNA\tNA\t471\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_imicola\t1.0\n",
      "GBDP4647-08\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_imicola\tNA\tAlgeria\tNA\tNA\tNA\tNA\t471\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_imicola\t1.0\n",
      "GBMIN46771-17\tArthropoda\tInsecta\tDiptera\tCulicidae\tCulicinae\tNA\tCulex\tCulex_pipiens\tCulex_pipiens_pipiens\tAlgeria\tNA\tNA\tNA\tNA\t450\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCulicidae\t1.0\tCulex\t1.0\tCulex_terzii\t0.95\n",
      "GBMIN54028-17\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_cataneii\tNA\tAlgeria\tNA\tNA\tNA\tNA\t472\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_cataneii\t1.0\n",
      "GBMIN54029-17\tArthropoda\tInsecta\tDiptera\tCeratopogonidae\tCeratopogoninae\tNA\tCulicoides\tCulicoides_cataneii\tNA\tAlgeria\tNA\tNA\tNA\tNA\t472\tcellularOrganisms\t1.0\tEukaryota\t1.0\tMetazoa\t1.0\tArthropoda\t1.0\tInsecta\t1.0\tDiptera\t1.0\tCeratopogonidae\t1.0\tCulicoides\t1.0\tCulicoides_cataneii\t1.0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "source ../../../code/classify.sh\n",
    "\n",
    "RDPcoiresults2tsv enafroCOI_all_clean_Diptera_taxa enafroCOI_diptera_all_data_taxa\n",
    "\n",
    "#sed -i 's/|l-/\\|/g; s/|fam-/\\|/g; s/|subfam-/\\|/g; s/|tri-/\\|/g; s/|gs-/\\|/g; s/|sp-/\\|/g; s/|subsp-/\\|/g; s/|country-/\\|/g; s/|exactsite-/\\|/g; s/|lat_/\\|/g; s/|lon_/\\|/g; s/|elev-/\\|/g; s/|/\\t/g; s/\\t\\t/\\t/g' enafroCOI_diptera_taxa\n",
    "#awk 'BEGIN {FS=\"\\t\";OFS=\"\\t\";}; BEGINFILE{\n",
    "#print \"Process_ID\",\"Phylum\",\"Class\",\"Order\", \"Family\",\"Subfamily\", \\\n",
    "#\"Tribe\",\"Genus\",\"Species\",\"Subspecies\", \\\n",
    "#\"Country\",\"Exactsite\",\"Latitude\",\"Longitude\",\"Elevation\",\"seq_len\", \\\n",
    "#\"As_CellularOrganism\",\"CellularOrganism_sc\",\"As_superkingdom\",\"Superkingdom_sc\", \\\n",
    "#\"As_kingdom\",\"Kingdom_sc\",\"As_phylum\",\"Phylum_sc\",\"As_class\",\"Class_sc\", \\\n",
    "#\"As_order\",\"Order_sc\",\"As_family\",\"Family_sc\",\"As_genus\",\"Genus_sc\", \\\n",
    "#\"As_species\",\"Species_sc\"};{\n",
    "#    print $1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16, \\\n",
    "#    $17,$19,$20,$22,$23,$25,$26,$28,$29,$31,$32,$34,$35,$37,$38,$40,$41,$43}' \\\n",
    "#    enafroCOI_diptera_taxa > enafroCOI_diptera_taxa.tsv\n",
    "head -10 enafroCOI_all_clean_Diptera_taxa.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **HEADERS problems**\n",
    "**Editting some problematic headers**  \n",
    "Some of the sequence headers lack the last field denoting the sequence length i.e \"l-***\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3436-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3436-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3437-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3437-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3438-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3438-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3439-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3439-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3440-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3440-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3448-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3448-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3472-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3472-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3473-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3473-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3474-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3474-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3502-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3502-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3503-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3503-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3504-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3504-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3505-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3505-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3506-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3506-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3507-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3507-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3617-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3617-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3618-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3618-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3619-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3619-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3620-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3620-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3621-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3621-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3647-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3647-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3648-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3648-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3649-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3649-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3650-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3650-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3651-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3651-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3686-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3686-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3687-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3687-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3688-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3688-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3689-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3689-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3690-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3690-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3691-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3691-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3692-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3692-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3693-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3693-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3694-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3694-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3695-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3695-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3702-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3702-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3703-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3703-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3704-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3704-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3705-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3705-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3706-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3706-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3728-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3728-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3729-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3729-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3730-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3730-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3731-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3731-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3732-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3732-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3734-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3734-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3735-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3735-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3736-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3736-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3737-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3737-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3738-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3738-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3739-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3739-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3740-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3740-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3741-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3741-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3742-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3742-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3743-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3743-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3746-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3746-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3747-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3747-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3748-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3748-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3749-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3749-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3907-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3907-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3914-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3914-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3917-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3917-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3918-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3918-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3919-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3919-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3920-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3920-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3921-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3921-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3937-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3937-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3940-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3940-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3964-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3964-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3965-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3965-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3966-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3966-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3967-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3967-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3968-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3968-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3969-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3969-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4100-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4100-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4101-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4101-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4102-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4102-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4109-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4109-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4110-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4110-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4111-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4111-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4112-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4112-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4113-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4113-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4135-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4135-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4137-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4137-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4138-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4138-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4139-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4139-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4155-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4155-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4174-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4174-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS4175-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS4175-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3785-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3785-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3786-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3786-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3787-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3787-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3788-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3788-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3789-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3789-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3790-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3790-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3791-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3791-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3792-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3792-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3793-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3793-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n",
      "To delete sequences with specific words in the headers please choose [Yes] to proceed or [No] to cancel\n",
      "\n",
      "Proceeding with infile.fasta...\n",
      "1 records match the pattern: KINS3794-13\n",
      "\tConcatinating sequence lines for each record in infile.fasta...\n",
      "\tRemoving any records with 'KINS3794-13' description in header from file...\n",
      "\n",
      "\tDONE. All deleted records have been stored in 'infile_unwanted.fasta'\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cat: cd: No such file or directory\n",
      "1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: 1) YES\n",
      "2) NO\n",
      "Select option YES to delete, [1] or NO to exit, [2]: Select option YES to delete, [1] or NO to exit, [2]: "
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'. ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\\ncd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\\n\\ncp enafroCOI_diptera_all_data.fasta infile.fasta\\nfor line in `cat cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera/process_IDs`\\ndo\\ndelete_unwanted infile.fasta << EOF\\n1\\n$line\\nEOF\\ndone\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-257a96a290b3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'bash'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'. ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\\ncd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\\n\\ncp enafroCOI_diptera_all_data.fasta infile.fasta\\nfor line in `cat cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera/process_IDs`\\ndo\\ndelete_unwanted infile.fasta << EOF\\n1\\n$line\\nEOF\\ndone\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2321\u001b[0m             \u001b[0mmagic_arg_s\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar_expand\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstack_depth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2322\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2323\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2324\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2325\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mnamed_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m                 \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshebang\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0;31m# write a basic docstring:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<decorator-gen-109>\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    243\u001b[0m             \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_error\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_script\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_close\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'. ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\\ncd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\\n\\ncp enafroCOI_diptera_all_data.fasta infile.fasta\\nfor line in `cat cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera/process_IDs`\\ndo\\ndelete_unwanted infile.fasta << EOF\\n1\\n$line\\nEOF\\ndone\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "\n",
    "cp enafroCOI_diptera_all_data.fasta infile.fasta\n",
    "for line in `cat cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera/process_IDs`\n",
    "do\n",
    "delete_unwanted infile.fasta << EOF\n",
    "1\n",
    "$line\n",
    "EOF\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tRemoving gaps, '-', from infile_unwanted.fasta\n",
      "\tConcatinating sequence lines for each record in infile_unwanted.fasta...\n",
      ">KINS3436-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3437-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3438-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3439-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3440-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_Ikuywa|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3448-13|Arthropoda|Insecta|Diptera|fam-Simuliidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka_in_the_fields|lat_0.212|lon_34.924|elev-1570.43|l-658\n",
      ">KINS3472-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-KEEP_Snake_park|lat_0.252|lon_34.92|elev-1614.17|l-658\n",
      ">KINS3473-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-KEEP_Snake_park|lat_0.252|lon_34.92|elev-1614.17|l-658\n",
      ">KINS3474-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-KEEP_Snake_park|lat_0.252|lon_34.92|elev-1614.17|l-658\n",
      ">KINS3502-13|Arthropoda|Insecta|Diptera|fam-Athericidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3503-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3504-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3505-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3506-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3507-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3617-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3618-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3619-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3620-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n",
      ">KINS3621-13|Arthropoda|Insecta|Diptera|fam-Chironomidae|subfam-NA|tri-NA|gs-NA|sp-NA|subsp-NA|country-Kenya|exactsite-Choka|lat_0.228|lon_34.963|elev-1643.49|l-658\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "#removing gaps. The output is \"input_dgpd.fasta\", without gaps, \"-\".\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "remove_gaps infile_unwanted.fasta\n",
    "#Introducing a field \"l-xxx\" that has the length of the sequence in the header\n",
    "awk '/^>/{hdr=$0; next}\n",
    "    { seq=$0 } match(seq,/^.*$/) { LEN=RLENGTH }\n",
    "    { print hdr\"|l-\"LEN; print seq }' infile_unwanted_dgpd.fasta > infile_unwanted_dgpd_edited.fasta\n",
    "\n",
    "mv infile_unwanted_dgpd_edited.fasta infile_unwanted.fasta\n",
    "rm infile_unwanted_dgpd.fasta\n",
    "grep \">\" infile_unwanted.fasta > infile_headers_fasta\n",
    "head -20 infile_headers_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ~/bioinformatics/github/co1_metaanalysis/data/input/diptera\n",
    "#substituting headers from infile_headers_fasta into all *diptera*.fasta files.\n",
    ". ~/bioinformatics/github/co1_metaanalysis/code/process_all_input_files.sh\n",
    "replacing_headers diptera_clean_headers_fasta *diptera*.fasta << EOF\n",
    "infile_headers_fasta\n",
    "EOF\n",
    "#cleaning up residual files\n",
    "mv infile_headers_fasta dipt99_headers_fasta\n",
    "rm -rf infile.fasta infile_unwanted.fasta outputfile process_IDs sedeW4xeJ"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
