# MANUSCRIPTS
From our last meeting with Dr Scott and Dr. Villinger (3rd of July 2019) the issue of publications that may come out of our work was discussed and the following options were reached upon:
1. A data release manuscript
2. A protocol/review manuscript
3. A data analysis outcome manuscript

The work and references for each is in the linked google drive folder
## Protocol/review paper
In this case a review is to be done on what has been the usage of the data we are analysing so far then:
### Questions.
a. What has been done with this data so far?
b. What more can be done from this data?
c. Is phylogenetic and phylogeographic analysis one possiblity?
d. How can it be done?
>(i) What are the challanges and how can they be solved?
>(ii) What is the result and how informative is it?
>(iii) How can this be improved on?

This way a phylogenetic and phylogeographic analysis is justified and a protocol paper given purpose.
This question can be adequately answered through the pipeline developemnt and testing currently approaching completion. The 279,205 records so far retrieved from the pupblic API has been key to this analysis and development
N/B: Thinking along the lines of: Big/BOLD data phylogenetic and comparative phylogeography analysis, the case of Barcode of Life Database Cytochrome c Oxidase subunit 1 data.

## Data release paper
Data from the KenBOL project funded by a Canadian grant under the iBOL project were never released/published. Thus this data may still be lying unfinished -proper annotation in some cases - and largely still kept unpublished by individual data managers.
Therefore there is a need to finish up on the data and publish it through a data release paper.
Examples of data release papers are given in the linked Google Drive folder.
Email from Dr. Scott about this:
> "----- Forwarded Message -----
From: Miller, Scott <MillerS@si.edu>
To: Gilbert Kibet (kibet.gilbert@ymail.com) <kibet.gilbert@ymail.com>
Cc: jandouwe@icipe.org <jandouwe@icipe.org>
Sent: Thursday, July 4, 2019, 09:27:24 AM GMT+3
Subject: Kenya insect data release paper?
>
>Regarding your need for a rapid publication from your masters project, you might do a data release paper for some or all of the insect data from the IDRC KenBOL project?
>As we discussed, because of the sudden termination of funding for the KenBOL project, most of the insect data were "abandoned" without being "finished" or "published".  Most of the data are not public on BOLD, so they are not benefiting the scientific community.  Thus, reviewing the quality of the sequences and the locality data and the identifications now, and publishing the data on BOLD and GenBank, would be a useful publication, and we have models for DNA barcode "data release" papers (see attached).  These data release papers are not great contributions to science, but they are solid publications, and some of them have become well cited because the data are popular for other uses.
>These papers can include various forms of comments on the data (as in the examples attached), but mostly about making the data available with high quality metadata, and not attempting detailed analysis.
>If you want to do this, you might work first with Bob Copeland on specimens that he provided.  If you give him printed trees, he can mark them up with current identifications (giving him printed trees will save both of you time versus having him look at trees online).  Then you could look at data provided by others.  Fabien Haas was an entomologist at icipe at the time, and provided aquatic insects and crop pests - he is in Germany now, and would probably be willing to help.  Others came from NMK, and Laban would be the best contact there.  How much of the KenBOL data you include would be up to you and the various collaborators.
If you update names and data in BOLD (which can be done individually through the web interface or in bulk by uploading spreadsheets), then it is easy to submit the records to GenBank through BOLD.  Some of the collaborators can update their own records in BOLD, and but it will be faster in some cases for you to get their comments and update the records yourself.
I think at least a solid core of the records could be reviewed and published quickly.  I think Bob would be an eager collaborator.  I will be happy to assist, especially with general methods and identifications of any Lepidoptera.
>--Scott"

**Challenges:**
The main Challenge as noted above by Scott is access to data. Most of which still private, but can be accessed by directly communicating with the data managers: the same also allude to in the email below:
>"----- Forwarded Message -----
From: Miller, Scott <MillerS@si.edu>
To: Dan Masiga <dmasiga@icipe.org>
Cc: Gilbert Kibet (kibet.gilbert@ymail.com) <kibet.gilbert@ymail.com>; jandouwe@icipe.org <jandouwe@icipe.org>; Miller, Scott <MillerS@si.edu>
Sent: Thursday, July 4, 2019, 08:59:53 AM GMT+3
Subject: KenBOL records
>
>I was reviewing Kibet Gilbert's project with him today, including the data sources he is using.  He has the KenBOL records from BOLD container IDRCK, but I noticed that project KLEPI (542 sequences of Lepidoptera) is missing.  I only have access to KLEPI, but it looks like it is in BOLD container KINST.  Can you please check KINST to make sure he has access to all the KenBOL records?
>Thanks, Scott"

## Final data analysis outcome paper
This will be the final paper that exhaustively reports on the outcome of our study. It will conclusively answer the research questions in the proposal i.e:
**General Objective:**
To  construct  a  series  of  phylogenetic  trees  based  on  COI  barcode  sequences  of arthropod specimens  hosted  by  BOLD  database  and  establish  the  phylogenetic distribution  of  these organisms within East Africa. And to conduct phylogeographic assessment based on their associated geographic coordinates and metadata.
**Specific Objectives:**
1. To identify phylogenetic and phylogeographic distribution of  fruit  flies (Tephritidae  family) and biting flies (Diptera: True flies) within East Africa.
2. To  identify   insect  species  that  are  potential  pests  or  disease   vectors  from combined phylogenetic similarity and phylogeographic distribution.
3. To identify any potential new classification or species if any.

**Research Questions:**
1. What is the phylogenetic diversity of phylum Arthropoda within the East Africa region based on COI barcode sequences available in BOLD database as compared to arthropod traditional taxonomical diversity already established based largely on ecological and morphological features?
2. How are the phylum Arthropoda organisms phylogeographically distributed within East Africa based on GIS coordinates of voucher specimen hosted in BOLD database and what does that imply about their evolution?

**Points to note:**
1. Nuclear mitochondrial (NUMT) DNA/pseudogenes:

>"From: Miller, Scott
Sent: Friday, June 1, 2018 1:51 PM
To: 'Gilbert Kibet' <kibet.gilbert@ymail.com>
Cc: Masiga, Daniel <dmasiga@icipe.org>; Steven Nyanjom <gernyanjom@yahoo.com>; Jandouwe Villinger <jandouwe@icipe.org>; Caleb Kipkurui <kipkuruis@hotmail.com>
Subject: RE: MSc of Molecular Biology and Bioinformatics Thesis -- meta analysis of East African arthropod barcodes
>
>The answer to question one is fitting the data to the question (or vice versa).  I would not use COI for phylogeny at the family level or above.  COI performs very well for analysis at the species and generic level for almost all insect groups.
>You can use several techniques to combine family level trees from other studies with your species and genus level trees from COI.  You can also selectively sequence more genes from exemplar specimens (using DNA residues from Guelph or taking new legs).  See for example what we did in the attached study in Papua New Guinea, which built on a COI base.
>When DNA barcoding started, there was a lot of concern about NUMTs, but they have not turned out to be a serious problem for insects.  The only insect group where they turn up routinely is Orthoptera, and it seems that they are fairly easy to recognize and filter out.  BOLD has quality control tools that would catch most of them.  But you will want to review the sequences that you use for possible NUMTs. ..."
>
>"----- Forwarded Message -----
From: Miller, Scott <MillerS@si.edu>
To: Gilbert Kibet <kibet.gilbert@ymail.com>
Cc: Masiga, Daniel <dmasiga@icipe.org>; Steven Nyanjom <gernyanjom@yahoo.com>; Jandouwe Villinger <jandouwe@icipe.org>; Caleb Kipkurui <kipkuruis@hotmail.com>
Sent: Saturday, June 2, 2018, 05:24:00 PM GMT+3
Subject: RE: MSc of Molecular Biology and Bioinformatics Thesis -- meta analysis of East African arthropod barcodes
>
>I just saw this citation, which reminded me that I should be clear that while Orthoptera seem to be the only insect order in which NUMTs regularly show up, NUMTs do occasionally show up elsewhere in the insects.  So you do want to check for themâ€¦
>Genome Biol Evol. 2017 Oct 1;9(10):2732-2738. doi: 10.1093/gbe/evx173.
>The Trouble with MEAM2: Implications of Pseudogenes on Species Delimitation in the Globally Invasive Bemisia tabaci (Hemiptera: Aleyrodidae) Cryptic Species Complex."
